src/expr/symbol_table.cpp

```diff
@@ -205,29 +229,62 @@ bool OverloadedTypeTrie::markOverloaded(const string& name, Expr obj) {
 
   // types can be identical but vary on the kind of the type, thus we must
   // distinguish based on this
-  std::map<Type, Expr>::iterator it = tat->d_symbols.find(rangeType);
-  if (it != tat->d_symbols.end()) {
-    Expr prev_obj = it->second;
-    // if there is already an active function with the same name and expects the
-    // same argument types
-    if (isOverloadedFunction(prev_obj)) {
-      if (prev_obj.getType() == obj.getType()) {
-        // types are identical, simply ignore it
-        return true;
-      } else {
-        // otherwise there is no way to distinguish these types, we return an
-        // error
+  if (d_allowFunctionVariants || argTypes.empty())
```

Comment:
So, if I understand correctly, when the flag is true, we allow functions to have the same name and argument types but to return different types.  We also always allow functions/constants with no arguments to have different types.  However, we never allow the redefinition of a function or constant with the same name, same arguments, and same return type as an existing definition.

When the flag is false, we don't allow function definitions with the same name and argument types, even if they return a different type.

Right?

Reply:
Yes that's right. I've added further documentation in my latest commit to make this clear.